lib/chef/provisioning/aws_driver/driver.rb

```diff
@@ -101,23 +115,83 @@ def allocate_load_balancer(action_handler, lb_spec, lb_options, machine_specs)
-        enable_zones = (availability_zones || []).dup
-        disable_zones = []
-        actual_elb.availability_zones.each do |availability_zone|
-          if !enable_zones.delete(availability_zone.name)
-            disable_zones << availability_zone.name
+        # TODO: refactor this whole giant method into many smaller method calls
+        # Update scheme - scheme is immutable once set, so if it is changing we need to delete the old
+        # ELB and create a new one
+        if scheme.downcase != actual_elb.scheme
+          desc = ["  updating scheme to #{scheme}"]
+          desc << "  WARN: scheme is immutable, so deleting and re-creating the ELB"
+          perform_action.call(desc) do
+            actual_elb.delete
+            actual_elb = elb.load_balancers.create(lb_spec.name, lb_optionals)
           end
         end
-        if enable_zones.size > 0
-          perform_action.call("  enable availability zones #{enable_zones.join(', ')}") do
-            actual_elb.availability_zones.enable(*enable_zones)
+
+        # Update security groups
+        if security_group_ids.empty?
+            Chef::Log.info("No Security Groups specified in resource; load_balancer[#{actual_elb.name}] cannot have " +
+            "empty Security Groups, so assuming it only currently has the default Security Group.  No action taken.")
+        else
+          current = Set.new(actual_elb.security_group_ids)
+          desired = Set.new(security_group_ids)
+          if current != desired
+            perform_action.call("  updating security groups to #{desired.to_a}") do
+              elb.client.apply_security_groups_to_load_balancer(
+                load_balancer_name: actual_elb.name,
+                security_groups: desired.to_a
+              )
+            end
           end
         end
-        if disable_zones.size > 0
-          perform_action.call("  disable availability zones #{disable_zones.join(', ')}") do
-            actual_elb.availability_zones.disable(*disable_zones)
+
+        # You can only have 1 of either subnets or availability zones, and if in a VPC you must have
+        # at least one of either.
+        if !availability_zones.empty?
+
+          enable_zones = (availability_zones || []).dup
+          disable_zones = []
+          actual_elb.availability_zones.each do |availability_zone|
+            if !enable_zones.delete(availability_zone.name)
+              disable_zones << availability_zone.name
+            end
+          end
+          if enable_zones.size > 0
+            perform_action.call("  enable availability zones #{enable_zones.join(', ')}") do
+              actual_elb.availability_zones.enable(*enable_zones)
+            end
           end
+          if disable_zones.size > 0
+            perform_action.call("  disable availability zones #{disable_zones.join(', ')}") do
+              actual_elb.availability_zones.disable(*disable_zones)
+            end
+          end
+
+        elsif !subnets.empty?
```

Comment:
Thats why I have this check.  If you start with availability zones and switch to subnets, it simply tries to update the subnets.  If you start with subnets and switch to availability zones, it tries to update the availability zones.

If you specify both, it will update first availability zones then subnets - this will possibly cause the resource to converge every time, but I was going to chalk that up to a user error.

If you specify neither then neither the availability zones or subnets change.  I did this because it isn't possible for a load balancer to belong to 0 availability zones (and consequently, it also cannot belong to 0 subnets).